### PR TITLE
Fix massive memory leak

### PR DESCRIPTION
**Pull request by asottile**

### CODE REVIEW

**Comment by bukzor** [comment]
LGTM. Waiting for @eevee's say.

**Comment by asottile** [comment]
Failing build: https://travis-ci.org/asottile/cheetah/builds/20298367
Succeeding build: https://travis-ci.org/asottile/cheetah/builds/20298526

File: cheetah/Tests/NameMapper.py [hunk 1]
```diff
@@ -540,6 +544,49 @@ def test_int(self):
         self.assertEquals(5, t.intify('5'))
 
 
+class SearchListObject(object):
+    class foo(object):
+        class bar(object):
+            baz = object()
```
Comment (eevee) [inline cheetah/Tests/NameMapper.py]: would also like to see a test for an object that has itself as a property, since that's an easy edge case to overlook

i.e. an object such that `foo.bar is foo`

File: cheetah/c/cheetah.h [hunk 2]
```diff
@@ -63,13 +63,16 @@
...
-    if ( PyNamemapper_hasKey(nameSpace, nameChunks[0]) ) {\
+    if (PyNamemapper_hasKey(nameSpace, nameChunks[0]) ) {\
 		instrumentRecordNameSpaceIndex(placeholderID, currentSearchListIndex);\
         /* We always do the first lookup with useDottedNotation = 1, because
          * otherwise looking up locals and globals (from the dicts locals() and
          * globals()) would always fail. */ \
-        theValue = PyNamemapper_valueForName(nameSpace, &nameChunks[0], 1, placeholderID, executeCallables, 1); \
+        theValue_tmp = theValue = PyNamemapper_valueForName(nameSpace, &nameChunks[0], 1, placeholderID, executeCallables, 1); \
         theValue = PyNamemapper_valueForName(theValue, &nameChunks[1], numChunks - 1, placeholderID, executeCallables, useDottedNotation);\
+        if (theValue != theValue_tmp) {\
```
Comment (eevee) [inline cheetah/c/cheetah.h]: yes and this approach will fail the test i just proposed.

this should instead look like

```
        theValue_tmp = theValue = PyNamemapper_valueForName(nameSpace, &nameChunks[0], 1, placeholderID, executeCallables, 1); \
        if (numChunks > 1) { \
            theValue = PyNamemapper_valueForName(theValue_tmp, &nameChunks[1], numChunks - 1, placeholderID, executeCallables, useDottedNotation);\
            Py_DECREF(theValue_tmp);\
        }\
```

File: cheetah/Tests/NameMapper.py (hunk 1 above)
Comment (bukzor) [inline cheetah/Tests/NameMapper.py]: I feel like the current tests cover that case sufficiently; it's just a special case of what's already covered.

Can you explain the scenario you're worried about?

File: cheetah/c/cheetah.h (hunk 2 above)
Comment (bukzor) [inline cheetah/c/cheetah.h]: I don't believe that's true.

I don't understand your logic around `numChunks > 1`, but given that `PyNamemapper_valueForName` works like it should, it won't incref when the input is the output.

Certainly I could be wrong, but I'd like to know why.

Comment (eevee) [inline cheetah/c/cheetah.h]: the second call could be in or outside the `if`, since it's a no-op when it gets a count of zero; putting it inside just makes it slightly clearer that we need the decref _because_ we're making a second call that actually does something.

if you look up `foo.bar` where `foo.bar is bar`, the following happens with your patch
1. just `foo` is looked up by the first call.  it's been INCREF'd, because you have to INCREF a return value, obviously.
2. `.bar` is looked up by the second call.  it's been INCREF'd again.  but it's really the same object, `foo`, now INCREF'd twice.
3. `theValue == theValue_tmp`, because they're exactly the same object, so your check does not DECREF.

this will _not_ happen if you merely look up `foo`, because it's the second call that causes the extra INCREF with no corresponding DECREF in the first place.

i don't understand why you wrote it this way in the first place tbh; whether there are any chunks left to evaluate is a much more direct measure of whether the second call does anything.

incidentally: should we be checking for `NULL` as well, in the event that the first call raises an exception?

**Comment by eevee** [comment]
:shipit:
